### ISSUE TITLE
GPU automatic memory defrag

### ISSUE THREAD

**Issue by lukemetz** [google/jax#9064]
Currently the TFRT TPU backend of supports automatic memory defragmentation.

It would be great if other platforms, in particular GPU, could also implement this functionality.

@zhangqiaorjc says this can be implemented as device neutral, slow defrag by basically copying device memory to host and then copying it back with appropriate synchronizations.

**Comment by zhangqiaorjc** [comment]
@lukemetz do you have a sample program that can tell us (XLA/GPU) a bit more about the kinds of workload that cause fragmentation today? They have mitigations if the workloads are in certain ways.

cc @jurahul @cheshire @sanjoy

**Comment by zhangqiaorjc** [comment]
cc @skye @hawkinsp

**Comment by lukemetz** [comment]
Hey sorry for the delay -- didn't have GH notifications on.

The types of workloads are almost always when there is more than one jax function.

I think it is caused something like (though I just spent a bit trying to get this to fail and couldn't.... so there might be something more nuanced going on.):
```
losses = []
for i in range(100):
  loss = some_function(data)
  losses.append(loss)
```

In this particular example, I can just cast loss to a np array (or cast losses[-2] to a np array). 

Sometimes this is nested inside some other computation though. Ideally I would like to keep the entire thing Jax.

The exact program I had in mind when I filed this was a variant on: https://github.com/google/learned_optimization/blob/main/learned_optimization/examples/simple_lopt_train.py though when I tried to reproduce this just now the OOM seems gone? Did anything change on the jax side?

**Comment by zhangqiaorjc** [comment]
@lukemetz if this is no longer affecting you, we can probably lower the priority for now
